telegram/src/main/kotlin/com/github/kotlintelegrambot/dispatcher/handlers/InputChainHandler.kt

```diff
@@ -0,0 +1,65 @@
+package com.github.kotlintelegrambot.dispatcher.handlers
+
+import com.github.kotlintelegrambot.Bot
+import com.github.kotlintelegrambot.entities.Message
+import com.github.kotlintelegrambot.entities.Update
+import com.github.kotlintelegrambot.extensions.filters.Filter
+
+data class InputChainHandlerEnvironment<T>(
+    val inputChainIdGen: (previousInputChainId: T?, Update) -> T?,
+    val inputChainId: T,
+    val bot: Bot,
+    val update: Update,
+    val message: Message,
+)
+
+class InputChainHandler<T>(
+    private val filter: Filter,
+    private val initialHandler: Handler,
+    private val inputChainIdGen: (previousInputChainId: T?, Update) -> T?,
+    private val handleInputChain: HandleInputChain<T>,
+) : Handler {
+    private val chainMap = mutableMapOf<Long, T?>()
+
+    override fun checkUpdate(update: Update): Boolean =
+        initialHandler.checkUpdate(update) || update.check()
+
+    override suspend fun handleUpdate(bot: Bot, update: Update) {
+        update.message?.let { message ->
+            message.from?.id?.let { userId ->
```

Comment:
~~Given that `handleUpdate` is guaranteed to be called after `check` returns `true` we normally make some assumptions (using assertions) about pre-conditions here (for example `message`, `from` and `id` are not null) even if the compiler can't guarantee them~~

~~In the future it might be good to look into wheter we can leverage kotlin contracts to improve this a bit.~~

Nevermind, I realised we can't do that here because `checkUpdate` also checks for initial handler's check which does not guarantee any of these things.